pkg/reconciler/timeout_handler.go

```diff
@@ -0,0 +1,232 @@
+package reconciler
+
+import (
+	"sync"
+
+	"time"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	clientset "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	defaultFunc = func(i interface{}) {}
+)
+
+const (
+	defaultTimeout = 10 * time.Minute
+)
+
+// StatusKey interface to be implemented by Taskrun Pipelinerun types
+type StatusKey interface {
+	GetRunKey() string
+}
+
+// TimeoutSet contains required k8s interfaces to handle build timeouts
+type TimeoutSet struct {
+	logger                  *zap.SugaredLogger
+	kubeclientset           kubernetes.Interface
+	pipelineclientset       clientset.Interface
+	taskRunCallbackFunc     func(interface{})
+	pipelineRunCallbackFunc func(interface{})
+	stopCh                  <-chan struct{}
+	statusMap               *sync.Map
+	done                    map[string]chan bool
+	doneMut                 sync.Mutex
+}
+
+// NewTimeoutHandler returns TimeoutSet filled structure
+func NewTimeoutHandler(
+	kubeclientset kubernetes.Interface,
+	pipelineclientset clientset.Interface,
+	stopCh <-chan struct{},
+	logger *zap.SugaredLogger,
+) *TimeoutSet {
+	return &TimeoutSet{
+		kubeclientset:     kubeclientset,
+		pipelineclientset: pipelineclientset,
+		stopCh:            stopCh,
+		statusMap:         &sync.Map{},
+		done:              make(map[string]chan bool),
+		doneMut:           sync.Mutex{},
+		logger:            logger,
+	}
+}
+
+// SetTaskRunCallbackFunc sets the callback function when timeout occurs for taskrun objects
+func (t *TimeoutSet) SetTaskRunCallbackFunc(f func(interface{})) {
+	t.taskRunCallbackFunc = f
+}
+
+// SetPipelineRunCallbackFunc sets the callback function when timeout occurs for pipelinerun objects
+func (t *TimeoutSet) SetPipelineRunCallbackFunc(f func(interface{})) {
+	t.pipelineRunCallbackFunc = f
+}
+
+// Release function deletes key from timeout map
+func (t *TimeoutSet) Release(runObj StatusKey) {
+	key := runObj.GetRunKey()
+	t.doneMut.Lock()
+	defer t.statusMap.Delete(key)
+	defer t.doneMut.Unlock()
+
+	if finished, ok := t.done[key]; ok {
+		delete(t.done, key)
+		close(finished)
+	}
+}
+
+func (t *TimeoutSet) getOrCreateFinishedChan(runObj StatusKey) chan bool {
+	var finished chan bool
+	key := runObj.GetRunKey()
+	t.doneMut.Lock()
+	if existingfinishedChan, ok := t.done[key]; ok {
+		finished = existingfinishedChan
+	} else {
+		finished = make(chan bool)
+	}
+	t.done[key] = finished
+	t.doneMut.Unlock()
```

Comment:
so minor: curious why not `defer t.doneMut.Unlock()` in this function?

Reply:
Behaviour remains the same either way.

Reply:
ah kk, just a bit inconsistent since we're doing `defer` above, no big deal tho!